File: app/src/main/java/org/ole/planet/myplanet/ui/team/teamMember/AdapterMemberRequest.java
package org.ole.planet.myplanet.ui.team.teamMember;

import android.content.Context;
import android.view.LayoutInflater;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import org.ole.planet.myplanet.databinding.RowMemberRequestBinding;
import org.ole.planet.myplanet.model.RealmMyTeam;
import org.ole.planet.myplanet.model.RealmUserModel;

import java.util.List;

import io.realm.Realm;

public class AdapterMemberRequest extends RecyclerView.Adapter<AdapterMemberRequest.ViewHolderUser> {
    private RowMemberRequestBinding rowMemberRequestBinding;
    private Context context;
    private List<RealmUserModel> list;
    private Realm mRealm;
    private String teamId;

    public AdapterMemberRequest(Context context, List<RealmUserModel> list, Realm mRealm) {
        this.context = context;
        this.list = list;
        this.mRealm = mRealm;
    }

    public void setTeamId(String teamId) {
        this.teamId = teamId;
    }

    @NonNull
    @Override
    public ViewHolderUser onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        rowMemberRequestBinding = RowMemberRequestBinding.inflate(LayoutInflater.from(context), parent, false);
        return new ViewHolderUser(rowMemberRequestBinding);
    }

    @Override
    public void onBindViewHolder(@NonNull ViewHolderUser holder, int position) {
        if (list.get(position).toString().equals(" ")) {
            rowMemberRequestBinding.tvName.setText(list.get(position).name);
        } else {
            rowMemberRequestBinding.tvName.setText(list.get(position).toString());
        }
        rowMemberRequestBinding.btnAccept.setOnClickListener(view -> {
            acceptReject(list.get(position), true, position);

        });
        rowMemberRequestBinding.btnReject.setOnClickListener(view -> acceptReject(list.get(position), false, position));
    }

    private void acceptReject(RealmUserModel userModel, boolean isAccept, int position) {
        if (!mRealm.isInTransaction()) mRealm.beginTransaction();
        RealmMyTeam team = mRealm.where(RealmMyTeam.class).equalTo("teamId", teamId).equalTo("userId", userModel.id).findFirst();
        if (team != null) {
            if (isAccept) {
                team.setDocType("membership");
                team.setUpdated(true);
            } else {
                team.deleteFromRealm();
            }
        }
        mRealm.commitTransaction();
        list.remove(position);
        notifyDataSetChanged();
    }

    @Override
    public int getItemCount() {
        return list.size();
    }

    static class ViewHolderUser extends RecyclerView.ViewHolder {
        RowMemberRequestBinding rowMemberRequestBinding;

        public ViewHolderUser(RowMemberRequestBinding rowMemberRequestBinding) {
            super(rowMemberRequestBinding.getRoot());
            this.rowMemberRequestBinding = rowMemberRequestBinding;
        }
    }
}

File: app/src/main/java/org/ole/planet/myplanet/ui/team/teamMember/AdapterMemberRequest.kt
package org.ole.planet.myplanet.ui.team.teamMember

import android.content.Context
import android.view.LayoutInflater
import android.view.ViewGroup
import androidx.recyclerview.widget.RecyclerView
import org.ole.planet.myplanet.R
import org.ole.planet.myplanet.databinding.RowMemberRequestBinding
import org.ole.planet.myplanet.model.RealmUserModel
import org.ole.planet.myplanet.utilities.DiffUtils
import org.ole.planet.myplanet.utilities.Utilities

import androidx.recyclerview.widget.ListAdapter

class AdapterMemberRequest(
    private val context: Context,
    private val currentUser: RealmUserModel,
    private val onAction: (RealmUserModel, Boolean) -> Unit
) : ListAdapter<RealmUserModel, AdapterMemberRequest.ViewHolderUser>(MWC_DIFF_CALLBACK) {
    companion object {
        val MWC_DIFF_CALLBACK = DiffUtils.itemCallback<RealmUserModel>(
            areItemsTheSame = { oldItem, newItem -> oldItem.id == newItem.id },
            areContentsTheSame = { oldItem, newItem -> oldItem.name == newItem.name }
        )
    }
    private var teamId: String? = null
    private var teamLeader = false
    private var joinedTeamMembers = 0

    fun setTeamId(teamId: String?) {
        this.teamId = teamId
    }

    fun setData(members: List<RealmUserModel>, isLeader: Boolean, memberCount: Int) {
        teamLeader = isLeader
        joinedTeamMembers = memberCount
        submitList(members)
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolderUser {
        val binding = RowMemberRequestBinding.inflate(LayoutInflater.from(parent.context), parent, false)
        return ViewHolderUser(binding)
    }

    override fun onBindViewHolder(holder: ViewHolderUser, position: Int) {
        val currentItem = getItem(position)
        val binding = holder.binding
        binding.tvName.text = currentItem.name ?: currentItem.toString()

        with(binding) {
            val userCanModerateRequests = teamLeader
            val isRequester = currentItem.id == currentUser.id
            btnAccept.isEnabled = joinedTeamMembers < 12
            btnReject.isEnabled = true
            btnAccept.setOnClickListener(null)
            btnReject.setOnClickListener(null)

            if (isRequester) {
                btnAccept.isEnabled = false
                btnReject.isEnabled = false
                btnAccept.setOnClickListener(null)
                btnReject.setOnClickListener(null)
            } else if (isGuestUser() || !userCanModerateRequests) {
                btnAccept.isEnabled = false
                btnReject.isEnabled = false
            } else {
                btnAccept.setOnClickListener { handleClick(holder, true) }
                btnReject.setOnClickListener { handleClick(holder, false) }
            }
        }
    }

    private fun isGuestUser() = currentUser.id?.startsWith("guest") == true

    private fun handleClick(holder: RecyclerView.ViewHolder, isAccepted: Boolean) {
        val adapterPosition = holder.bindingAdapterPosition
        if (adapterPosition != RecyclerView.NO_POSITION) {
            val targetUser = getItem(adapterPosition)
            if (targetUser.id == currentUser.id) return
            onAction(targetUser, isAccepted)
        }
    }

    class ViewHolderUser(val binding: RowMemberRequestBinding) : RecyclerView.ViewHolder(binding.root)
}
